File: labs/mobile/donor-archive/tunnel/UniversalMeshEvasionPipeline.kt
package com.luminet.android.tunnel

enum class AndroidPipelineTier {
    DIRECT_MESH,
    HOLE_PUNCHED_MESH,
    DPI_EVADED_MESH,
    STEALTH_BRIDGE_FALLBACK
}

data class AndroidPipelineMetrics(
    var packetsProcessed: Long = 0,
    var modeSwitches: Int = 0,
    var currentTier: AndroidPipelineTier = AndroidPipelineTier.DIRECT_MESH
)

class UniversalMeshEvasionPipeline(val peerId: String) {
    var consecutiveErrors: Int = 0
        private set
    val metrics = AndroidPipelineMetrics()

    val escalationThreshold: Int = 3

    fun processOutboundFrame(frame: ByteArray): List<ByteArray> {
        metrics.packetsProcessed++
        return when (metrics.currentTier) {
            AndroidPipelineTier.DIRECT_MESH, AndroidPipelineTier.HOLE_PUNCHED_MESH -> {
                listOf(frame)
            }
            AndroidPipelineTier.DPI_EVADED_MESH -> {
                if (frame.size > 8) {
                    val mid = frame.size / 2
                    listOf(frame.copyOfRange(0, mid), frame.copyOfRange(mid, frame.size))
                } else {
                    listOf(frame)
                }
            }
            AndroidPipelineTier.STEALTH_BRIDGE_FALLBACK -> {
                val prefix = "STH:".toByteArray(Charsets.UTF_8)
                val wrapped = ByteArray(prefix.size + frame.size)
                System.arraycopy(prefix, 0, wrapped, 0, prefix.size)
                System.arraycopy(frame, 0, wrapped, prefix.size, frame.size)
                listOf(wrapped)
            }
        }
    }

    fun reportFailure(): AndroidPipelineTier {
        consecutiveErrors++
        if (consecutiveErrors >= escalationThreshold) {
            consecutiveErrors = 0
            val nextTier = when (metrics.currentTier) {
                AndroidPipelineTier.DIRECT_MESH -> AndroidPipelineTier.HOLE_PUNCHED_MESH
                AndroidPipelineTier.HOLE_PUNCHED_MESH -> AndroidPipelineTier.DPI_EVADED_MESH
                AndroidPipelineTier.DPI_EVADED_MESH -> AndroidPipelineTier.STEALTH_BRIDGE_FALLBACK
                AndroidPipelineTier.STEALTH_BRIDGE_FALLBACK -> AndroidPipelineTier.STEALTH_BRIDGE_FALLBACK
            }

            if (nextTier != metrics.currentTier) {
                metrics.currentTier = nextTier
                metrics.modeSwitches++
            }
        }
        return metrics.currentTier
    }

    fun reportSuccess() {
        consecutiveErrors = 0
    }
}

File: labs/control-ui/api-shadow/universal_mesh_evasion_pipeline.ts
// Control UI TypeScript: Universal Mesh Evasion Pipeline (Batch 10 Synthesis C1)

export type PipelineTransportTier =
  | 'direct_mesh'
  | 'hole_punched_mesh'
  | 'dpi_evaded_mesh'
  | 'stealth_bridge_fallback';

export interface PipelineMetrics {
  packetsProcessed: number;
  modeSwitches: number;
  currentTier: PipelineTransportTier;
}

export class UniversalMeshEvasionPipeline {
  public peerId: string;
  public consecutiveErrors: number = 0;
  public readonly escalationThreshold: number = 3;
  public metrics: PipelineMetrics = {
    packetsProcessed: 0,
    modeSwitches: 0,
    currentTier: 'direct_mesh',
  };

  constructor(peerId: string) {
    this.peerId = peerId;
  }

  public processOutboundFrame(frame: Uint8Array): Uint8Array[] {
    this.metrics.packetsProcessed++;

    switch (this.metrics.currentTier) {
      case 'direct_mesh':
      case 'hole_punched_mesh':
        return [frame];

      case 'dpi_evaded_mesh': {
        if (frame.length > 8) {
          const mid = Math.floor(frame.length / 2);
          return [frame.slice(0, mid), frame.slice(mid)];
        }
        return [frame];
      }

      case 'stealth_bridge_fallback': {
        const prefix = new TextEncoder().encode('STH:');
        const wrapped = new Uint8Array(prefix.length + frame.length);
        wrapped.set(prefix, 0);
        wrapped.set(frame, prefix.length);
        return [wrapped];
      }
    }
  }

  public reportFailure(): PipelineTransportTier {
    this.consecutiveErrors++;
    if (this.consecutiveErrors >= this.escalationThreshold) {
      this.consecutiveErrors = 0;
      let nextTier: PipelineTransportTier = this.metrics.currentTier;

      if (this.metrics.currentTier === 'direct_mesh') {
        nextTier = 'hole_punched_mesh';
      } else if (this.metrics.currentTier === 'hole_punched_mesh') {
        nextTier = 'dpi_evaded_mesh';
      } else if (this.metrics.currentTier === 'dpi_evaded_mesh') {
        nextTier = 'stealth_bridge_fallback';
      }

      if (nextTier !== this.metrics.currentTier) {
        this.metrics.currentTier = nextTier;
        this.metrics.modeSwitches++;
      }
    }
    return this.metrics.currentTier;
  }

  public reportSuccess(): void {
    this.consecutiveErrors = 0;
  }
}
